b2share/modules/records/serializers/schemas/json.py

```diff
@@ -43,19 +47,35 @@ class DraftSchemaJSONV1(Schema):
     @pre_dump
     def filter_internal(self, data):
         """Remove internal fields from the record metadata."""
+        external_pids = []
+        bucket = None
+        try:
+            record = g.record
+            external_pids = generate_external_pids(record)
+            bucket = Bucket.query.filter_by(
+                id=str(record.files.bucket)).first()
+        except AttributeError:
+            record = None
+
         if '_deposit' in data['metadata']:
-            if 'external_pids' in data['metadata']['_deposit']:
-                data['metadata']['external_pids'] = \
-                    data['metadata']['_deposit']['external_pids']
             data['metadata']['owners'] = data['metadata']['_deposit']['owners']
+            # Add the external_pids only if the
+            # user is allowed to read the bucket
+            if external_pids and record and record.files:
+                if (allow_public_file_metadata(record) or
+                        files_permission_factory(bucket, 'bucket-read').can()):
+                    data['metadata']['external_pids'] = external_pids
             del data['metadata']['_deposit']
         if '_files' in data['metadata']:
-            if allow_public_file_metadata(data['metadata']):
+            # Also add the files field only if the user is allowed
+            if allow_public_file_metadata(data['metadata']) or \
+                    (bucket and files_permission_factory(
+                        bucket, 'bucket-read').can()):
                 data['files'] = data['metadata']['_files']
                 for _file in data['files']:
                     if 'external_pids' in data['metadata'] and \
                             any(d['key'] == _file['key']
-                                for d in data['metadata']['external_pids']):
+                                for d in external_pids):
                         _file['b2safe'] = True
```

Comment:
@thijscobben Is the field `files->b2safe` used by the B2Safe script? Otherwise we could rename it "external" or something more appropriate. With the current code we will have `files->b2safe=true` even for files which are not stored in B2SAFE (ex: TSD use case).